Posts/2021/09/my-crescendo-journey.md

```diff
@@ -0,0 +1,158 @@
+next step.
+
+## Writing the parser functions
+
+To create the parser functions I had to know what the output looked like for all of the possible
+command combinations of `vssadmin.exe`. I looked at the help provided by `vssadmin` and captured the
+output for each subcommand in a separate file. I used these output files to design and implement a
+parsing function for each subcommand.
+
+Now, on to the configuration file.
+
+## Creating the JSON configuration
+
+For this I used the example from the [blog post][blog3] as a template. I also looked at the
+`Get-InstalledPackage` example from the [Preview 2][blog2] blog post to see how the native commands
+were referenced. For my first cmdlet I started with this JSON configuration:
+
+```json
+{
+    "$schema": "./Microsoft.PowerShell.Crescendo.Schema.json",
+    "Commands": [
+         {
+            "Verb": "Get",
+            "Noun": "VssProvider",
+            "OriginalName": "$env:Windir/system32/vssadmin.exe",
+            "OriginalCommandElements": [
+                "list",
+                "providers"
+            ],
+            "OutputHandlers": [
+                {
+                    "ParameterSetName": "Default",
+                    "HandlerType": "Function",
+                    "Handler": "ParseProvider"
+                }
+            ],
+         }
+    ]
+}
+```
+
+The `ParseProvider` function is one of the functions that I had written to parse the output. I
+repeated this pattern to create a new cmdlet for each of the `vssadmin` subcommands.
+
+Notice that the first line of the JSON references a schema file. This file comes with the Crescendo
+module. I copied that to the folder where I was developing my module. I used Visual Studio Code (VS
+Code) to do all my development. With this schema file, VS Code provides IntelliSense for the JSON,
+making it easy to know what values are required and the type of information needed.
+
+Eventually, I added properties to the configuration for full help with descriptions and examples.
+And I defined parameter sets for the `vssadmin` commands that supported parameters.
+
+## Creating the new module
+
+Crescendo, itself, is a module. It contains cmdlets that help you create your configuration and then
+uses that configuration to create the module containing your cmdlets. Once I was happy with the
+configuration file, I used the `Export-CrescendoModule` cmdlet to create my module.
+
+```powershell
+Export-CrescendoModule -ConfigurationFile .\vssadmin.crescendo.config.json -ModuleName VssAdmin.psm1
+```
+
+Crescendo created two new files:
+
+- The module code file `VssAdmin.psm1`
+- The module manifest file `VssAdmin.psd1`
+
+These are the only two files that need to be installed. The `VssAdmin.psm1` file contains all the
+cmdlets that Crescendo generated from the configuration and the **Output Handler** functions I
+wrote to parse the output into objects.
+
```

Comment:
not sure if you want to add that script output handler is put in the output directory next to the module files.

Reply:
Will add this now that I know about it.

Reply:
I decided this was out of scope for this blog post. We need to document it. I will talk about when I present this to the user group. Also, I may do a follow up post that examines the configuration and parsing functions in more detail. That will give me a chance to talk about the trade-offs and design decisions.

I like the script output handlers because you can make runtime changes without export and re-import. But I like the function-based handlers to keep the module down to two files.